### ISSUE TITLE
TEEC_RequestCancellation not working as expected..! 

### ISSUE THREAD

**Issue by kumarabekur** [OP-TEE/optee_os#1895]
Hello there,

I am using Hello world : https://github.com/linaro-swg/hello_world example to understand working of TEEC_RequestCancellation API. I have created a thread to cancel the TA. In that thread I am cancelling TA after a small delay. I have given more delay in TA so that TA is still alive when TEEC_RequestCancellation using TEE_Wait(). [Here](https://gist.github.com/kumarabekur/fbbbc13f443269e97a63c897debb1e31) is the gist containing HOST, TA and output.

I have used TEE_UnmaskCancellation, TEE_GetCancellationFlag APIs. I have referred TEE_Internal_Core_API_Specification and TEE_Client_API_Specification document. Please let me know if I am missing something. Any idea @jenswi-linaro @jbech-linaro @jforissier @vchong @MrVan  Also I would like to know more about working of TEEC_RequestCancellation.

Thanks in Advance. :)

**Comment by jenswi-linaro** [comment]
The output looks OK to me. Keep in mind that the kernel may schedule threads as it likes.

**Comment by jforissier** [comment]
@jenswi-linaro this line looks suspicious:
```
DEBUG:   USER-TA:inc_value:131: CANCEL FLAG =0
```
I would expect flag to be 1, especially since there is a reasonable (?) time for the cancellation status to reach the TEE (1 second). But as you said, subject to scheduling.

Now, I looked at the GP test suite, there is a cancellation test (d3-ee-b1  which is gp_7038 in xtest). The interesting thing is, this test is commented out by default, and if I enable it it fails (tested on QEMU):
```
* gp_7038 d3-ee-b1
/home/jerome/work/optee_repo_qemu/optee_test/host/xtest/gp_7000.c:764: TEEC_InvokeCommand(SESSION01, 0x00000004, OPERATION01, &ret_orig) has an unexpected value: 0xffff0000, expected 0xffff0002
  gp_7038 FAILED
```

I *think* it is the same issue, because the body of the TA does this:
```C
   for(i = 0; i < 100; i++)      // 100 loops of 100ms = 10s
   {
      TEE_MaskCancellation();
      TEE_Wait(100);
      TEE_UnmaskCancellation();
      if(TEE_GetCancellationFlag() == true)
         return TEE_ERROR_CANCEL;
   }
   /* if the flag was never set */
   return TEE_ERROR_GENERIC;
```
So, if `TEE_GetCancellationFlag()` never returns `true`, the function returns `TEE_ERROR_GENERIC` (0xffff0000) which is what I observed.

**Comment by jenswi-linaro** [comment]
I'm reading the log as `TEEC_RequestCancellation()` is called after `TEEC_OpenSession()` has returned but before `TEEC_InvokeCommand()` has been called.

Perhaps regression case `1009` should be updated to include something with `TEE_MaskCancellation()` and `TEE_UnmaskCancellation()`.

**Comment by jforissier** [comment]
> I'm reading the log as `TEEC_RequestCancellation()` is called after `TEEC_OpenSession()` has returned but before `TEEC_InvokeCommand()` has been called.

Exactly, so `TA_InvokeCommandEntryPoint()`, called because of `TEEC_InvokeCommand()`, should detect that cancellation has been requested and thus `TEE_GetCancellationFlag()` should return 1, not 0 (at line 131).

**Comment by jenswi-linaro** [comment]
A cancellation request can't be queued, it's only performed on running invokes or queued invokes. Here's there's nothing active in secure world when `TEEC_RequestCancellation()` is called.

**Comment by jforissier** [comment]
@jenswi-linaro ah, right, my mistake. Sorry, that was not obvious to me when I read the spec.

That leaves us with the `xtest 7038` failure, which does mask/unmask, but that's a different story.

**Comment by kumarabekur** [comment]
@jforissier unfortunately I cant able to find the code snippet you showed in my xtest source code. I am using 2.4 version of OPTEE OS/Client and xtest . 
@jforissier , @jenswi-linaro 
As per Client Spec doc and Internal API doc, 
**Client Side :**  the `started `of `TEEC_Operation` should be made 0.
                     `TEEC_Operation `struct should be passed in OpenSession
                      The same `TEEC_Operation` struct should not be passed to another `Opensession` or `InvokeCommand `  (This is taken care in my sample app by giving larger sleep (1 sec more) before calling `TEEC_RequestCancellation`    
**TA side :** By default all cancellation requests are blocked  so   `TEE_GetCancellationFlag()` returning 0 in line 131 is expected. But After calling `TEE_UnmaskCancellation()` (Returns 1 :expected ) I am again checking `TEE_GetCancellationFlag()` (line 137)  . I am expecting it to return 1. But this is not the case. Is there anything other I am missing in Client side or TA side which is causing this behaviour. Since I have called  `TEEC_RequestCancellation`  after 5 sec in a thread I am expecting TA to return immediately after 5sec (There is 10 sec delay in TA). But this is not the case.  :(

**Comment by jforissier** [comment]
@kumarabekur `xtest 7038` is a GlobalPlatform test and is not part of the public test suite.

**Comment by kumarabekur** [comment]
Thank you @jforissier . :) Any idea about issue I am facing ?? :) It is weird that it is not working as documented. :( just wondering if I have missed some trick here

**Comment by kumarabekur** [comment]
I have also tried removing `sleep 6` . In that case I have to pass different `TEEC_Operation` struct because as per Client spec doc : 

> Client Applications MUST NOT reuse the Operation structure for another Operation until the cancelled
>command has actually returned in the thread executing the _TEEC_OpenSession_ or
> _TEEC_InvokeCommand_ function.

So I don't have the option to use same `TEEC_Operation` struct . Here also I am cancelling the `TEEC_Operation` struct  which I have passed to `TEEC_OpenSession` . In this case which TEEC_Operation struct I should be using with `TEEC_RequestCancellation()` ?? Any suggestions would be really helpful :)

**Comment by jforissier** [comment]
It's OK to declare/define only one global `TEEC_Operation op;`. Use it first to open a session, then clear it, and re-use it (1) in the cancel thread and (2) to invoke a TA command.

FYI we have such a simple cancellation test in `xtest 1009` (https://github.com/OP-TEE/optee_test/blob/master/host/xtest/regression_1000.c#L711-L778). Maybe you can start from that, because it uses the API correctly. Then you may modify it to check the cancellation flag and/or mask/unmask cancellation in the TA. Hopefully things should be clear.

**Comment by kumarabekur** [comment]
In Client Spec doc it is mentioned that `TEEC_Operation op` should be passed to `OpenSession` with started member initialized to 0. i.e., we have to pass op to  `OpenSession` . But in https://github.com/OP-TEE/optee_test/blob/master/host/xtest/regression_1000.c#L711-L778 it is not been done..! : 
```
		if (ADBG_EXPECT_TEEC_SUCCESS(c,
		         xtest_teec_open_session(&session, &os_test_ta_uuid,
                          NULL, &ret_orig))) {
```
any reason for  this?? (L734-L735)

**Comment by jforissier** [comment]
`NULL` is valid, too.

`TEEC_OpenSession()`
...
**Parameters**
`operation`: a pointer to an Operation containing a set of Parameters to exchange with the
Trusted Application, or `NULL` if no Parameters are to be exchanged or if the operation cannot be cancelled.
...

**Comment by kumarabekur** [comment]
yes.  But if I want to cancel the TA then I guess we should pass op with `op.started = 0` not `NULL` . If we pass `NULL` then we cannot cancel the TA.

**Comment by jforissier** [comment]
You don't cancel a TA, you cancel an operation, which is either "open session" or "invoke command". By passing NULL to "open session" you're saying you don't want this particular operation to be cancel-able. This does not imply anything on the later "invoke command" operations.

**Comment by kumarabekur** [comment]
In code snippet you showed above : 
```
   for(i = 0; i < 100; i++)      // 100 loops of 100ms = 10s
   {
      TEE_MaskCancellation();
      TEE_Wait(100);
      TEE_UnmaskCancellation();
      if(TEE_GetCancellationFlag() == true)
         return TEE_ERROR_CANCEL;
   }
```
there is 100 loops of 100ms . There is no a single `TEE_Wait` with 10000ms(`TEE_Wait`(10000) ?? =10s) . This indicates TEE_Wait will return only after timeout so a very big dealy is not given in `TEE_wait`?? I got this doubt because the `TEEC_RequestCancellation` actually cancelling a TA now and my `InvokeCommand` fails with code 0xffff0002(_TEEC_ERROR_CANCEL_) . But it waits 10s before exiting .
